presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java

```diff
@@ -208,20 +273,50 @@ static long getEstimatedKllInMemorySize(@Nullable KllItemsSketch<?> sketch, Clas
 
     static SketchParameters<?> getSketchParameters(Type type)
     {
-        if (type.getJavaType().equals(double.class)) {
-            return new SketchParameters<>(Double::compareTo, new ArrayOfDoublesSerDe());
+        if (!type.isOrderable() || !type.isComparable()) {
+            throw new PrestoException(INVALID_ARGUMENTS, type + " does not support comparisons or ordering");
         }
-        else if (type.getJavaType().equals(long.class)) {
-            return new SketchParameters<>(Long::compareTo, new ArrayOfLongsSerDe());
+        // catch specific cases first (e.g. real and short decimal), moving to wider comparisons
+        // later
+        if (type.equals(REAL)) {
+            return new SketchParameters<>(Double::compareTo, new ArrayOfDoublesSerDe(),
+                    (Object intValue) -> (double) Float.intBitsToFloat(((Long) intValue).intValue()));
+        }
+        else if (isShortDecimal(type)) {
+            DecimalType decimalType = (DecimalType) type;
+            return new SketchParameters<>(Double::compareTo, new ArrayOfDoublesSerDe(),
+                    (Object longValue) -> new SqlDecimal(BigInteger.valueOf((long) longValue), decimalType.getScale(), decimalType.getScale()).toBigDecimal().doubleValue());
         }
-        else if (type.getJavaType().equals(Slice.class)) {
-            return new SketchParameters<>(String::compareTo, new ArrayOfStringsSerDe());
+        else if (isLongDecimal(type)) {
+            DecimalType decimalType = (DecimalType) type;
+            return new SketchParameters<>(Double::compareTo, new ArrayOfDoublesSerDe(),
+                    (Object encodedDecimal) -> new SqlDecimal(Decimals.decodeUnscaledValue((Slice) encodedDecimal), decimalType.getScale(), decimalType.getPrecision()).toBigDecimal().doubleValue());
```

Comment:
It does not, but neither does the optimizer when calculating min/max values on statistics. To implement this without precision loss, I'd need to implement the `ArrayOfItemsSerde` class from Apache Datasketches for `BigDecimal`, maybe called `ArrayOfBigDecimalSerde` or something. 

My issue with this is that the optimizer statistics only work on doubles currently, so it is losing precision already with decimal types. On top of that if we actually stored the BigDecimal representation, to properly query the sketch in the optimizer you would end up going from:

- Presto decimal -> BigDecimal -> double -> and convert back to BigDecimal to query the sketch. Since we already lose precision in the optimizer converting to double, I opted to go this route for now.

The other issue with doing this "properly" is that if someone were to consume the KLL sketch produced by Presto, in order to parse it they would need to declare a dependency on `presto-main` (or wherever I put the `ArrayOfBigDecimalSerde` in order to deserialize the sketch. I could also try to make that contribution to the datasketches repo and then upgrade the dependency on the next release, but that will draw out the timeline for histograms.

Reply:
While this function is intended for the optimizer, it's also a public function and may be used on its own.  This seems like a correctness bug when taken out of the context of usage within the optimizer.  Would it vastly affects statistics gathering if we just don't allow it for big decimal types for now?

Reply:
Many filters on TPC-DS queries are on decimal types. It is quite important that we support them

Reply:
Another proposal is that we create a hidden function, similar to `sum_data_size_for_stats` and call is `kll_sketch_for_stats`. This version can be hidden and support decimals, while the `sketch_kll` we leave out the decimal support.

Reply:
We used to encode long decimals as binary blobs, before switching them to two long integers.  Could we serialize them into something that is backed by a byte array, such as a String?

Reply:
The problem is with the fact that it needs to be a type for which `ArrayOfItemsSerde` is implemented so that it can be inserted into the sketch. Currently we have `Double`, `Long`, and `String`. There is no option for `BigDecimal`

Reply:
For continuity: after some offline discussion, we will remove decimal support from this function and incorporate a change into the statistics portion of the SPI to execute some kind of project against the input argument to an aggregation function. This way we can leave out decimal support here and just cast the inputs for decimal-type columns to doubles